types/wlr_drm_lease_v1.c

```diff
@@ -0,0 +1,663 @@
+	}
+
+	int fd = wlr_drm_create_lease(outputs, request->n_connectors,
+			&lease->lessee_id);
+	if (fd < 0) {
+		wlr_log_errno(WLR_ERROR, "drm_create_lease failed");
+		wp_drm_lease_v1_send_finished(lease->resource);
+		return NULL;
+	}
+
+	lease->connectors = calloc(request->n_connectors,
+			sizeof(struct wlr_drm_lease_connector_v1 *));
+	if (!lease->connectors) {
+		wlr_log(WLR_ERROR, "Failed to allocate lease connectors list");
+		wp_drm_lease_v1_send_finished(lease->resource);
+		return NULL;
+	}
+	lease->n_connectors = request->n_connectors;
+	for (size_t i = 0; i < request->n_connectors; ++i) {
+		lease->connectors[i] = request->connectors[i];
+		lease->connectors[i]->active_lease = lease;
+	}
+
+	wlr_log(WLR_DEBUG, "Granting request %p", request);
+
+	wp_drm_lease_v1_send_lease_fd(lease->resource, fd);
+	close(fd);
+
+	return lease;
+}
+
+void wlr_drm_lease_request_v1_reject(
+		struct wlr_drm_lease_request_v1 *request) {
+	assert(request && request->lease);
+
+	wlr_log(WLR_DEBUG, "Rejecting request %p", request);
+
+	request->invalid = true;
+	wp_drm_lease_v1_send_finished(request->lease->resource);
+}
+
+void wlr_drm_lease_v1_revoke(struct wlr_drm_lease_v1 *lease) {
+	assert(lease);
+	wlr_log(WLR_DEBUG, "Revoking lease %"PRIu32, lease->lessee_id);
+
+	wp_drm_lease_v1_send_finished(lease->resource);
+
+	struct wlr_drm_lease_device_v1 *device = lease->device;
+	wlr_drm_backend_terminate_lease(device->backend, lease->lessee_id);
+	lease->lessee_id = 0;
+
+	for (size_t i = 0; i < lease->n_connectors; ++i) {
+		lease->connectors[i]->active_lease = NULL;
+	}
+
+	drm_lease_v1_destroy(lease);
+}
+
+static void drm_lease_v1_handle_resource_destroy(struct wl_resource *resource) {
+	struct wlr_drm_lease_v1 *lease = drm_lease_v1_from_resource(resource);
+	drm_lease_v1_destroy(lease);
+}
+
+static void drm_lease_v1_handle_destroy(
+		struct wl_client *client, struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static struct wp_drm_lease_v1_interface lease_impl = {
+	.destroy = drm_lease_v1_handle_destroy,
+};
+
+static void drm_lease_request_v1_handle_resource_destroy(
+		struct wl_resource *resource) {
+	struct wlr_drm_lease_request_v1 *request =
+		drm_lease_request_v1_from_resource(resource);
+	drm_lease_request_v1_destroy(request);
+}
+
+static void drm_lease_request_v1_handle_request_connector(
+		struct wl_client *client, struct wl_resource *request_resource,
+		struct wl_resource *connector_resource) {
+	struct wlr_drm_lease_request_v1 *request =
+		drm_lease_request_v1_from_resource(request_resource);
+	struct wlr_drm_lease_connector_v1 *connector =
+		drm_lease_connector_v1_from_resource(connector_resource);
+
+	if (!connector || connector->active_lease) {
+		/* This connector offer has been withdrawn or is leased */
+		wlr_log(WLR_ERROR, "Failed to request connector");
+		request->invalid = true;
+		return;
+	}
+
+	wlr_log(WLR_DEBUG, "Requesting connector %s", connector->output->name);
```

Comment:
This doesn't check whether the connector has already been added. We should avoid adding the same connector multiple times to the list, otherwise a client can blow up our stack with the VLA we allocate in `wlr_drm_lease_request_v1_grant`.